docs/architecture/adr-009-ABCI-design.md

```diff
@@ -0,0 +1,260 @@
+To facilitate adoption and cross-language compatibility in the near-term, Amino
+v1 will:
+
+- be fully compatible with the subset of Protobuf3 that excludes `oneof`
+- use the Amino prefix system to provide interface types, as opposed to `oneof`
+  style union types.
+
+That said, an Amino v2 will be worked on to improve the performance of the
+format and its useability in cryptographic applications.
+
+
+### PubKey
+
+Encoding schemes infect software. As a generic middleware, ABCI aims to have
+some cross scheme compatibility. For this it has no choice but to include opaque
+bytes from time to time. While we will not enforce Amino encoding for these
+bytes yet, we need to provide a type system. The simplest way to do this is to
+use a type string.
+
+PubKey will now look like:
+
+```
+message PubKey {
+    string type
+    bytes data
+}
+```
+
+where `type` can be:
+
+- "ed225519", with `data = <raw 32-byte pubkey>`
+- "secp256k1", with `data = <33-byte OpenSSL compressed pubkey>`
+
+and generated types. At the least we should use a reflection-based protobuf so
+we can just encode our own types, rather than using protobuf generated ones.
+
+### Addresses
+
+To simplify and improve computing addresses, we change it to the first 20-bytes of the SHA256
+of the raw 32-byte public key.
+
+We continue to use the Bitcoin address scheme for secp256k1 keys.
+
+### Validators
+
+Add a `bytes address` field:
+
+```
+message Validator {
+    bytes address
+    PubKey pub_key
+    int64 power
+}
+```
+
+### RequestBeginBlock and AbsentValidators
+
+To simplify this, RequestBeginBlock will include the complete validator set,
+including the address, and voting power of each validator, along
+with a boolean for whether or not they voted:
+
+```
+message RequestBeginBlock {
+  bytes hash
+  Header header
+  LastCommitInfo last_commit_info
+  repeated Evidence byzantine_validators
+}
+
+message LastCommitInfo {
+  int32 CommitRound
+  repeated SigningValidator validators
+}
+
+message SigningValidator {
+    Validator validator
+    bool signed_last_block
+}
+```
+
+Note that in Validators in RequestBeginBlock, we DO NOT include public keys. Public keys are
+larger than addresses and in the future, with quantum computers, will be much
+larger. The overhead of passing them, especially during fast-sync, is
+significant.
+
+Additional, addresses are changing to be simpler to compute, further removing
+the need to include pubkeys here.
+
+In short, ABCI developers must be aware of both addresses and public keys.
+
+### ResponseEndBlock
+
+Since ResponseEndBlock includes Validator, it must now include their address.
+
+### InitChain
+
+Change RequestInitChain and ResponseInitChain to
+
+```
+message RequestInitChain {
+    int64 time
+    string chain_id
+    ConsensusParams consensus_params
+    repeated Validator validators
+    bytes app_state_bytes
+}
+
+message ResponseInitChain {
+    ConsensusParams consensus_params
```

Comment:
Could be helpful to expand on this field and clarify why we return it back to the caller and the content is.

Reply:
`Chain ResponseInitChain to allow the app to specify the initial validator set
and consensus parameters.` is that better ? ConsensusParams definition doesnt belong here